origin-growth/src/resources/authentication.js

```diff
@@ -0,0 +1,126 @@
+const _growthModels = require('../models')
+const _identityModels = require('origin-identity/src/models')
+const db = { ..._growthModels, ..._identityModels }
+const logger = require('../logger')
+const Web3 = require('web3')
+const enums = require('../enums')
+const crypto = require('crypto')
+
+const web3 = new Web3(process.env.PROVIDER_URL || 'http://localhost:8545')
+// TODO: have this stored somewhere in the db
+const currentAgreementMessage =
+  'I accept the terms of growth campaign version: 1.0'
+
+/**
+ * Authenticates user's enrollment to the growth campaign
+ * @param {string} referrer - Eth address of the referrer.
+ * @param {Array<string>>} recipients - List of email addresses.
+ */
+async function authenticateEnrollment(accountId, agreementMessage, signature) {
+  if (currentAgreementMessage !== agreementMessage) {
+    throw new Error(
+      `Incorrect agreementMessage. Expected: "${currentAgreementMessage}" received: "${agreementMessage}"`
+    )
+  }
+  const recoveredAccountId = web3.eth.accounts.recover(
+    agreementMessage,
+    signature
+  )
+
+  if (accountId !== recoveredAccountId) {
+    throw new Error('Recovered and provided accounts do not match')
+  }
+
+  const participant = await db.GrowthParticipant.findOne({
+    where: {
+      ethAddress: accountId
+    }
+  })
+
+  if (
+    participant !== null &&
+    participant.status === enums.GrowthParticipantStatuses.Banned
+  ) {
+    logger.warn(`Banned user: ${accountId} tried to enroll`)
+    throw new Error('This user is banned')
+  }
+
+  const authToken =
+    participant === null
+      ? crypto.randomBytes(64).toString('hex')
+      : participant.authToken
+
+  const participantData = {
+    ethAddress: accountId,
+    status: enums.GrowthParticipantStatuses.Active,
+    agreementId: agreementMessage,
```

Comment:
Should we have an growth_agreement table where the agreement text is stored and an id that is then referred to here ? One problem would be how to handle translation...
An alternative would be to have the user only signing an agreement ID (ex: "Campaign Agreement v1.0") which would not be translated, but it does look a bit trustworthy to user if we show them they are signing the agreement data ? We can chat and decide what's best...

Reply:
Yeah I have a `TODO` comment on a couple of places in the Dapp2 saying this should be read from db. Yeah lets talke about it today on our call.  Have made an agenda item [in our doc](https://docs.google.com/document/d/15nzudW56hQVIWHc7DRJI5XzxenM0-oc2p7z4vOqc8ms/edit?ts=5b7b483f)